# Proceso de Extracción, Transformación y Carga (ETL)

## Importar librerías

In [42]:
import pandas as pd
import utils

import warnings
warnings.simplefilter("ignore")

In [43]:
# Instalar esta librería en caso de no tenerla, para trabajar con excel
#pip install --upgrade pandas openpyxl

## Extracción de los datos

In [44]:
archivo = '..\BBDD_Hospitalización.xlsx'
df = pd.read_excel(archivo, engine='openpyxl')
print(f'Se cargaron los datos del archivo {archivo}')

Se cargaron los datos del archivo ..\BBDD_Hospitalización.xlsx


## Transformación

En primer lugar se exploran los primeros y los últimos registros para visualizar su estructura, así como obtener información sobre las columnas y los tipos de datos.

In [45]:
df

EDAD DIABETES HOSPITALIZACIÓN ULTIMO MES    PSA BIOPSIAS PREVIAS  \
0    53.0       NO                         NO   4.00               NO   
1    56.0       NO                         NO   7.70               NO   
2    57.0       NO                         NO  29.00               SI   
3    56.0       NO                         NO   7.00               NO   
4    55.0       NO                         NO  29.00               SI   
..    ...      ...                        ...    ...              ...   
565  69.0       NO                         NO   4.43              NaN   
566  67.0       NO                         NO   8.00               NO   
567  59.0       NO                         NO   9.00               SI   
568   NaN      NaN                        NaN    NaN              NaN   
569   NaN      NaN                        NaN    NaN              NaN   

    VOLUMEN PROSTATICO ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS  \
0                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
1                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
2                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
3                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
4                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
..                 ...                                     ...   
565                NaN          FLUOROQUINOLONA_AMINOGLICOSIDO   
566                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
567                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
568                NaN                                     NaN   
569                NaN                                     NaN   

     NUMERO DE MUESTRAS TOMADAS  CUP ENF. CRONICA PULMONAR OBSTRUCTIVA  \
0                          12.0   NO                                NO   
1                          12.0   NO                                NO   
2                          24.0   NO                                NO   
3                          12.0   NO                                NO   
4                          24.0   NO                                NO   
..                          ...  ...                               ...   
565                        20.0  NaN                                NO   
566                        12.0   NO                               NaN   
567                        12.0   NO                               NaN   
568                         NaN  NaN                               NaN   
569                         NaN  NaN                               NaN   

                       BIOPSIA  \
0                          NEG   
1                          NEG   
2                          NEG   
3                          NEG   
4                          NEG   
..                         ...   
565                        NEG   
566  ADENOCARCINOMA GLEASON 7    
567                        NEG   
568                        NaN   
569                        NaN   

    NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA  \
0                                                    1                          
1                                                    1                          
2                                                    1                          
3                                                    1                          
4                                                    1                          
..                                                 ...                          
565                                                 NO                          
566                                                 NO                          
567                                                 NO                          
568                                                NaN                          
569                                                NaN                          

    FIEBRE  ITU TIPO DE CULTIVO AGENTE AISLADO  \
0       SI   N

In [46]:
# Columnas del dataframe
df.columns

Index(['EDAD', 'DIABETES', 'HOSPITALIZACIÓN ULTIMO MES', 'PSA',
       'BIOPSIAS PREVIAS', 'VOLUMEN PROSTATICO',
       'ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS', 'NUMERO DE MUESTRAS TOMADAS',
       'CUP', 'ENF. CRONICA PULMONAR OBSTRUCTIVA', 'BIOPSIA',
       'NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA',
       'FIEBRE', 'ITU', 'TIPO DE CULTIVO', 'AGENTE AISLADO',
       'PATRON DE RESISTENCIA', 'HOSPITALIZACION', 'DIAS HOSPITALIZACION MQ',
       'DIAS HOSPITALIZACIÓN UPC'],
      dtype='object')

In [47]:
# Se filtran las filas duplicadas
df[df.duplicated(keep=False)]

EDAD DIABETES HOSPITALIZACIÓN ULTIMO MES   PSA BIOPSIAS PREVIAS  \
30   73.0       NO                         NO   7.5               NO   
32   73.0       NO                         NO   7.5               NO   
36   62.0       NO                         NO  16.0               NO   
37   69.0       NO                         NO  17.0               SI   
39   66.0       NO                         NO   6.4               NO   
..    ...      ...                        ...   ...              ...   
557  59.0       NO                         NO  16.0               NO   
560  57.0       NO                         NO   4.8               NO   
562  78.0       NO                         NO   9.3               NO   
568   NaN      NaN                        NaN   NaN              NaN   
569   NaN      NaN                        NaN   NaN              NaN   

    VOLUMEN PROSTATICO ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS  \
30                  SI            CEFALOSPORINA_AMINOGLUCOCIDO   
32                  SI            CEFALOSPORINA_AMINOGLUCOCIDO   
36                  SI                           OROQUINOLONAS   
37                  SI            CEFALOSPORINA_AMINOGLUCOCIDO   
39                  SI                                   OTROS   
..                 ...                                     ...   
557                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
560                 NO                                   OTROS   
562                 SI            CEFALOSPORINA_AMINOGLUCOCIDO   
568                NaN                                     NaN   
569                NaN                                     NaN   

     NUMERO DE MUESTRAS TOMADAS  CUP ENF. CRONICA PULMONAR OBSTRUCTIVA  \
30                         12.0   NO                          SI, ASMA   
32                         12.0   NO                          SI, ASMA   
36                         12.0   NO                                SI   
37                         12.0   NO                                SI   
39                         12.0   NO                                SI   
..                          ...  ...                               ...   
557                        24.0   NO                                NO   
560                        12.0   NO                                NO   
562                        12.0   NO                                NO   
568                         NaN  NaN                               NaN   
569                         NaN  NaN                               NaN   

                       BIOPSIA  \
30                         NEG   
32                         NEG   
36   ADENOCARCINOMA GLEASON 8    
37                         NEG   
39                         NEG   
..                         ...   
557                        NEG   
560                        NEG   
562                        NEG   
568                        NaN   
569                        NaN   

    NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA  \
30                                                  NO                          
32                                                  NO                          
36                                                  NO                          
37                                                  NO                          
39                                                  NO                          
..                                                 ...                          
557                                                 NO                          
560                                                 NO                          
562                                                 NO                          
568                                                NaN                          
569                                                NaN                          

    FIEBRE  ITU TIPO DE CULTIVO AGENTE AISLADO PATRON DE RESISTENCIA  \
30  

In [48]:
# Se eliminan las filas duplicadas en todas las columnas manteniendo la primera aparición
df = df.drop_duplicates(subset=df.columns, keep='first')
df

EDAD DIABETES HOSPITALIZACIÓN ULTIMO MES    PSA BIOPSIAS PREVIAS  \
0    53.0       NO                         NO   4.00               NO   
1    56.0       NO                         NO   7.70               NO   
2    57.0       NO                         NO  29.00               SI   
3    56.0       NO                         NO   7.00               NO   
4    55.0       NO                         NO  29.00               SI   
..    ...      ...                        ...    ...              ...   
564  64.0       NO                         NO   4.80               NO   
565  69.0       NO                         NO   4.43              NaN   
566  67.0       NO                         NO   8.00               NO   
567  59.0       NO                         NO   9.00               SI   
568   NaN      NaN                        NaN    NaN              NaN   

    VOLUMEN PROSTATICO ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS  \
0                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
1                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
2                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
3                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
4                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
..                 ...                                     ...   
564                 SI                                   OTROS   
565                NaN          FLUOROQUINOLONA_AMINOGLICOSIDO   
566                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
567                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
568                NaN                                     NaN   

     NUMERO DE MUESTRAS TOMADAS  CUP ENF. CRONICA PULMONAR OBSTRUCTIVA  \
0                          12.0   NO                                NO   
1                          12.0   NO                                NO   
2                          24.0   NO                                NO   
3                          12.0   NO                                NO   
4                          24.0   NO                                NO   
..                          ...  ...                               ...   
564                        12.0   NO                                NO   
565                        20.0  NaN                                NO   
566                        12.0   NO                               NaN   
567                        12.0   NO                               NaN   
568                         NaN  NaN                               NaN   

                       BIOPSIA  \
0                          NEG   
1                          NEG   
2                          NEG   
3                          NEG   
4                          NEG   
..                         ...   
564  ADENOCARCINOMA GLEASON 7    
565                        NEG   
566  ADENOCARCINOMA GLEASON 7    
567                        NEG   
568                        NaN   

    NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA  \
0                                                    1                          
1                                                    1                          
2                                                    1                          
3                                                    1                          
4                                                    1                          
..                                                 ...                          
564                                                 NO                          
565                                                 NO                          
566                                                 NO                          
567                                                 NO                          
568                                                NaN                          

    FIEBRE  ITU TIPO DE CULTIVO AGENTE AISLADO  \
0       SI   N

In [49]:
# Se verifica que ya no hay filas duplicadas
df[df.duplicated(keep=False)]

Empty DataFrame
Columns: [EDAD, DIABETES, HOSPITALIZACIÓN ULTIMO MES, PSA, BIOPSIAS PREVIAS, VOLUMEN PROSTATICO, ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS, NUMERO DE MUESTRAS TOMADAS, CUP, ENF. CRONICA PULMONAR OBSTRUCTIVA, BIOPSIA, NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA, FIEBRE, ITU, TIPO DE CULTIVO, AGENTE AISLADO, PATRON DE RESISTENCIA, HOSPITALIZACION, DIAS HOSPITALIZACION MQ, DIAS HOSPITALIZACIÓN UPC]
Index: []

In [50]:
# se resetean los indexs
df = df.reset_index(drop=True)

In [51]:
# Se observan los nulos y tipo de datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 461 entries, 0 to 460
Data columns (total 20 columns):
 #   Column                                                                     Non-Null Count  Dtype  
---  ------                                                                     --------------  -----  
 0   EDAD                                                                       460 non-null    float64
 1   DIABETES                                                                   460 non-null    object 
 2   HOSPITALIZACIÓN ULTIMO MES                                                 460 non-null    object 
 3   PSA                                                                        456 non-null    float64
 4   BIOPSIAS PREVIAS                                                           458 non-null    object 
 5   VOLUMEN PROSTATICO                                                         459 non-null    object 
 6   ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS                   

Se cuenta con 461 registros, cuyos indices van del 0 al 460, y 20 columnas. Muchos de ellos del tipo objeto y se observan algunas columnas con valores nulos por lo que se aplicará una función que verifica el contenido dentro de cada columna y hace un conteo de la cantidad de nulos en la misma.

In [52]:
utils.verificar_tipo_datos(df)

nombre_campo  \
0                                                EDAD   
1                                            DIABETES   
2                          HOSPITALIZACIÓN ULTIMO MES   
3                                                 PSA   
4                                    BIOPSIAS PREVIAS   
5                                  VOLUMEN PROSTATICO   
6             ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS   
7                          NUMERO DE MUESTRAS TOMADAS   
8                                                 CUP   
9                   ENF. CRONICA PULMONAR OBSTRUCTIVA   
10                                            BIOPSIA   
11  NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA...   
12                                             FIEBRE   
13                                                ITU   
14                                    TIPO DE CULTIVO   
15                                     AGENTE AISLADO   
16                              PATRON DE RESISTENCIA   
17                                    HOSPITALIZACION   
18                            DIAS HOSPITALIZACION MQ   
19                           DIAS HOSPITALIZACIÓN UPC   

                                         tipo_datos  no_nulos_%  nulos_%  \
0                                 [<class 'float'>]       99.78     0.22   
1                  [<class 'str'>, <class 'float'>]       99.78     0.22   
2                  [<class 'str'>, <class 'float'>]       99.78     0.22   
3                                 [<class 'float'>]       98.92     1.08   
4                  [<class 'str'>, <class 'float'>]       99.35     0.65   
5                  [<class 'str'>, <class 'float'>]       99.57     0.43   
6                  [<class 'str'>, <class 'float'>]       99.78     0.22   
7                                 [<class 'float'>]       99.78     0.22   
8                  [<class 'str'>, <class 'float'>]       99.57     0.43   
9                  [<class 'str'>, <class 'float'>]       99.35     0.65   
10                 [<class 'str'>, <class 'float'>]       99.78     0.22   
11  [<class 'int'>, <class 'str'>, <class 'float'>]       99.78     0.22   
12                 [<class 'str'>, <class 'float'>]       99.78     0.22   
13                 [<class 'str'>, <class 'float'>]       99.78     0.22   
14                 [<class 'str'>, <class 'float'>]       99.78     0.22   
15                 [<class 'str'>, <class 'float'>]       96.10     3.90   
16                 [<class 'str'>, <class 'float'>]       99.78     0.22   
17                 [<class 'str'>, <class 'float'>]       99.13     0.87   
18                                [<class 'float'>]       99.78     0.22   
19                                [<class 'float'>]       99.78     0.22   

    nulos  
0       1  
1       1  
2       1  
3       5  
4       3  
5       2  
6       1  
7       1  
8       2  
9       3  
10      1  
11      1  
12      1  
13      1  
14      1  
15     18  
16      1  
17      4  
18      1  
19      1

En principio se observa que todas las columnas tienen como mínimo 1 valor nulo y que la máxima cantidad de nulos es en la columna AGENTE AISLADO, con 18 valores que representan el 3.9%.

In [53]:
# Filtrar filas con valores nulos
df[df.isna().any(axis=1)]

EDAD DIABETES HOSPITALIZACIÓN ULTIMO MES    PSA BIOPSIAS PREVIAS  \
9    44.0       NO                         NO    NaN               NO   
26   75.0       NO                         NO  25.00               NO   
31   63.0       NO                         NO  10.80               NO   
70   57.0       NO                         NO    NaN               NO   
86   58.0       NO                         NO   7.80              NaN   
93   70.0       NO                         NO   7.20               NO   
94   74.0       NO                         NO   5.80               NO   
97   60.0       NO                         NO    NaN               SI   
100  65.0       NO                         NO   5.97               NO   
103  61.0       NO                         NO   7.00               NO   
109  56.0       NO                         NO   4.00               NO   
110  55.0       NO                         NO  17.00               NO   
115  62.0       NO                         NO    NaN               NO   
120  74.0       NO                         NO   8.00               NO   
122  69.0       NO                         NO  13.40               SI   
123  63.0       NO                         NO  11.00               NO   
124  63.0       NO                         NO  21.30               SI   
132  64.0       NO                         NO  15.60               NO   
133  71.0       NO                         NO   5.40               NO   
140  58.0       NO                         NO  17.00               NO   
145  66.0       NO                         NO  11.50               NO   
150  66.0       NO                         NO   5.10               NO   
161  68.0       NO                         NO  15.00               NO   
164  69.0       NO                         NO   5.10               NO   
457  69.0       NO                         NO   4.43              NaN   
458  67.0       NO                         NO   8.00               NO   
459  59.0       NO                         NO   9.00               SI   
460   NaN      NaN                        NaN    NaN              NaN   

    VOLUMEN PROSTATICO ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS  \
9                   NO          FLUOROQUINOLONA_AMINOGLICOSIDO   
26                  SI          FLUOROQUINOLONA_AMINOGLICÓSIDO   
31                  NO          FLUOROQUINOLONA_AMINOGLICÓSIDO   
70                  NO          FLUOROQUINOLONA_AMINOGLICOSIDO   
86                  SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
93                  SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
94                  NO          FLUOROQUINOLONA_AMINOGLICOSIDO   
97                  SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
100                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
103                 NO          FLUOROQUINOLONA_AMINOGLICOSIDO   
109                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
110                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
115                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
120                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
122                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
123                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
124                 SI                                   OTROS   
132                 SI            CEFALOSPORINA_AMINOGLUCOCIDO   
133                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
140                 NO          FLUOROQUINOLONA_AMINOGLICOSIDO   
145                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
150                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
161                 NO          FLUOROQUINOLONA_AMINOGLICOSIDO   
164                 SI                           OROQUINOLONAS   
457                NaN          FLUOROQUINOLONA_AMINOGLICOSIDO   
458                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
459                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
460                NaN                 

In [54]:
cantidad_registros_nulos = len(df[df.isna().any(axis=1)])
porcentaje = round((cantidad_registros_nulos / len(df)) * 100,2)

print(f'Hay {cantidad_registros_nulos} registros con algún nulo, lo que representa el {porcentaje}% del total de registros')

Hay 28 registros con algún nulo, lo que representa el 6.07% del total de registros


Se tienen 28 registros donde hay al menos un valor nulo, esto representa el 6% de los registros disponibles, por lo que se decide borrarlos.

In [55]:
df = df.dropna()
df

EDAD DIABETES HOSPITALIZACIÓN ULTIMO MES   PSA BIOPSIAS PREVIAS  \
0    53.0       NO                         NO   4.0               NO   
1    56.0       NO                         NO   7.7               NO   
2    57.0       NO                         NO  29.0               SI   
3    56.0       NO                         NO   7.0               NO   
4    55.0       NO                         NO  29.0               SI   
..    ...      ...                        ...   ...              ...   
452  55.0       NO                         NO   5.0               NO   
453  70.0       SI                         NO  18.0               NO   
454  75.0       NO                         NO  75.0               NO   
455  67.0       NO                         NO   6.0               NO   
456  64.0       NO                         NO   4.8               NO   

    VOLUMEN PROSTATICO ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS  \
0                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
1                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
2                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
3                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
4                   SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
..                 ...                                     ...   
452                 NO          FLUOROQUINOLONA_AMINOGLICOSIDO   
453                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
454                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
455                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
456                 SI                                   OTROS   

     NUMERO DE MUESTRAS TOMADAS CUP ENF. CRONICA PULMONAR OBSTRUCTIVA  \
0                          12.0  NO                                NO   
1                          12.0  NO                                NO   
2                          24.0  NO                                NO   
3                          12.0  NO                                NO   
4                          24.0  NO                                NO   
..                          ...  ..                               ...   
452                        12.0  NO                                NO   
453                        12.0  NO                                NO   
454                        12.0  NO                                NO   
455                        12.0  NO                                NO   
456                        12.0  NO                                NO   

                       BIOPSIA  \
0                          NEG   
1                          NEG   
2                          NEG   
3                          NEG   
4                          NEG   
..                         ...   
452  ADENOCARCINOMA GLEASON 7    
453                        NEG   
454  ADENOCARCINOMA GLEASON 8    
455   ADENOCARCINOMA GLEASON 6   
456  ADENOCARCINOMA GLEASON 7    

    NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA  \
0                                                    1                          
1                                                    1                          
2                                                    1                          
3                                                    1                          
4                                                    1                          
..                                                 ...                          
452                                                 NO                          
453                                                 NO                          
454                                                 NO                          
455                                                 NO                          
456                                                 NO                          

    FIEBRE ITU TIPO DE CULTIVO AGENTE AISLADO  \
0       SI  NO              NO         

In [56]:
# Se verifica nuevamente la cantidad de nulos
utils.verificar_tipo_datos(df)

nombre_campo  \
0                                                EDAD   
1                                            DIABETES   
2                          HOSPITALIZACIÓN ULTIMO MES   
3                                                 PSA   
4                                    BIOPSIAS PREVIAS   
5                                  VOLUMEN PROSTATICO   
6             ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS   
7                          NUMERO DE MUESTRAS TOMADAS   
8                                                 CUP   
9                   ENF. CRONICA PULMONAR OBSTRUCTIVA   
10                                            BIOPSIA   
11  NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA...   
12                                             FIEBRE   
13                                                ITU   
14                                    TIPO DE CULTIVO   
15                                     AGENTE AISLADO   
16                              PATRON DE RESISTENCIA   
17                                    HOSPITALIZACION   
18                            DIAS HOSPITALIZACION MQ   
19                           DIAS HOSPITALIZACIÓN UPC   

                        tipo_datos  no_nulos_%  nulos_%  nulos  
0                [<class 'float'>]       100.0      0.0      0  
1                  [<class 'str'>]       100.0      0.0      0  
2                  [<class 'str'>]       100.0      0.0      0  
3                [<class 'float'>]       100.0      0.0      0  
4                  [<class 'str'>]       100.0      0.0      0  
5                  [<class 'str'>]       100.0      0.0      0  
6                  [<class 'str'>]       100.0      0.0      0  
7                [<class 'float'>]       100.0      0.0      0  
8                  [<class 'str'>]       100.0      0.0      0  
9                  [<class 'str'>]       100.0      0.0      0  
10                 [<class 'str'>]       100.0      0.0      0  
11  [<class 'int'>, <class 'str'>]       100.0      0.0      0  
12                 [<class 'str'>]       100.0      0.0      0  
13                 [<class 'str'>]       100.0      0.0      0  
14                 [<class 'str'>]       100.0      0.0      0  
15                 [<class 'str'>]       100.0      0.0      0  
16                 [<class 'str'>]       100.0      0.0      0  
17                 [<class 'str'>]       100.0      0.0      0  
18               [<class 'float'>]       100.0      0.0      0  
19               [<class 'float'>]       100.0      0.0      0

Queda una única columna con dos tipos de datos, se procede a analizarla.

In [57]:
columna_analizar = 'NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA'

In [58]:
# Es una columna de número enteros. Al tener un 'NO' es que aparecen dos tipos de datos.
df[columna_analizar].unique()

array([1, 2, 3, 5, 9, 'NO'], dtype=object)

In [59]:
# Contar cuantos valores de NO hay
total_NO = df[columna_analizar].value_counts().get('NO', 0)
total_filas = len(df)
porcentaje = round((total_NO / total_filas) * 100,2)

print(f'Hay un total de {total_filas} registros, de los cuales {total_NO} son "NO", lo que representa el {porcentaje}%')

Hay un total de 433 registros, de los cuales 408 son "NO", lo que representa el 94.23%


In [60]:
# Se analiza el valor 'NO' en el contexto de los demas datos.
df[df[columna_analizar] == 'NO']

EDAD DIABETES HOSPITALIZACIÓN ULTIMO MES   PSA BIOPSIAS PREVIAS  \
27   77.0       NO                         NO  91.8               NO   
28   75.0       NO                         NO  21.0               NO   
29   77.0       NO                         NO  98.0               NO   
30   73.0       NO                         NO   7.5               NO   
32   63.0       NO                         NO  10.5               NO   
..    ...      ...                        ...   ...              ...   
452  55.0       NO                         NO   5.0               NO   
453  70.0       SI                         NO  18.0               NO   
454  75.0       NO                         NO  75.0               NO   
455  67.0       NO                         NO   6.0               NO   
456  64.0       NO                         NO   4.8               NO   

    VOLUMEN PROSTATICO ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS  \
27                  SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
28                  SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
29                  SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
30                  SI            CEFALOSPORINA_AMINOGLUCOCIDO   
32                  NO          FLUOROQUINOLONA_AMINOGLICOSIDO   
..                 ...                                     ...   
452                 NO          FLUOROQUINOLONA_AMINOGLICOSIDO   
453                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
454                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
455                 SI          FLUOROQUINOLONA_AMINOGLICOSIDO   
456                 SI                                   OTROS   

     NUMERO DE MUESTRAS TOMADAS CUP ENF. CRONICA PULMONAR OBSTRUCTIVA  \
27                          6.0  NO                          SI, EPOC   
28                         12.0  NO                          SI, EPOC   
29                          6.0  NO                          SI, EPOC   
30                         12.0  NO                          SI, ASMA   
32                         12.0  NO                          SI, ASMA   
..                          ...  ..                               ...   
452                        12.0  NO                                NO   
453                        12.0  NO                                NO   
454                        12.0  NO                                NO   
455                        12.0  NO                                NO   
456                        12.0  NO                                NO   

                        BIOPSIA  \
27   ADENOCARCINOMA GLEASON 10    
28    ADENOCARCINOMA GLEASON 7    
29   ADENOCARCINOMA GLEASON 10    
30                          NEG   
32    ADENOCARCINOMA GLEASON 7    
..                          ...   
452   ADENOCARCINOMA GLEASON 7    
453                         NEG   
454   ADENOCARCINOMA GLEASON 8    
455    ADENOCARCINOMA GLEASON 6   
456   ADENOCARCINOMA GLEASON 7    

    NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA  \
27                                                  NO                          
28                                                  NO                          
29                                                  NO                          
30                                                  NO                          
32                                                  NO                          
..                                                 ...                          
452                                                 NO                          
453                                                 NO                          
454                                                 NO                          
455                                                 NO                          
456                                                 NO                          

    FIEBRE ITU TIPO DE CULTIVO AGENTE AISLADO PATRON DE RESISTENCIA  \
27   

Al analizar la cantidad de valores 'NO' que tiene la columna y comparando con la información de las otras columnas, se asume que NO significa que la persona no tuvo internación, por lo que se cambiará ese valor por 0 días.

In [61]:
# Se cambian los valores NO por 0
df[columna_analizar] = df[columna_analizar].replace('NO', 0)
# Se verifican los valores únicos y el tipo de dato
df[columna_analizar].unique(), df[columna_analizar].apply(type).unique()

(array([1, 2, 3, 5, 9, 0], dtype=int64), array([<class 'int'>], dtype=object))

Ahora, en cuanto a las columnas numéricas se cambia el tipo de datos por el tipo entero.

In [62]:
# Se cambian los siguientes tipo de datos
df['EDAD'] = df['EDAD'].astype(int)
df['NUMERO DE MUESTRAS TOMADAS'] = df['NUMERO DE MUESTRAS TOMADAS'].astype(int)
df['DIAS HOSPITALIZACION MQ'] = df['DIAS HOSPITALIZACION MQ'].astype(int)
df['DIAS HOSPITALIZACIÓN UPC'] = df['DIAS HOSPITALIZACIÓN UPC'].astype(int)

Se observan las variables categóricas de columnas para buscar unificar en caso de ser necesario.

In [63]:
# Se revisan las categorías de la columna 
df['ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS'].unique()

array(['FLUOROQUINOLONA_AMINOGLICOSIDO', 'CEFALOSPORINA_AMINOGLUCOCIDO',
       'OROQUINOLONAS', 'OTROS', 'FLUOROQUINOLONA_AMINOGLICÓSIDO'],
      dtype=object)

In [64]:
# Se renombran
df['ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS'] = df['ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS'].replace('FLUOROQUINOLONA_AMINOGLICÓSIDO', 
                                                                                                      'FLUOROQUINOLONA_AMINOGLICOSIDO')

In [65]:
# Se revisan las categorías de la columna 
df['ENF. CRONICA PULMONAR OBSTRUCTIVA'].unique()

array(['NO', 'SI, EPOC', 'SI, ASMA', 'SI'], dtype=object)

In [66]:
# Se renombran
df['ENF. CRONICA PULMONAR OBSTRUCTIVA'] = df['ENF. CRONICA PULMONAR OBSTRUCTIVA'].replace('SI, EPOC', 
                                                                                          'SI')
df['ENF. CRONICA PULMONAR OBSTRUCTIVA'] = df['ENF. CRONICA PULMONAR OBSTRUCTIVA'].replace('SI, ASMA', 
                                                                                          'SI')

In [67]:
# Se revisan las categorías de la columna 
df['BIOPSIA'].unique()

array(['NEG', 'ADENOCARCINOMA GLEASON 6 ', 'ADENOCARCINOMA GLEASON 7 ',
       'ADENOCARCINOMA GLEASON 6', 'ADENOCARCINOMA GLEASON 10 ',
       'ADENOCARCINOMA GLEASON 9 ', 'ADENOCARCINOMA GLEASON 8 ',
       'PROSTATITIS', 'ADENOCARCINOMA GLEASON 7',
       'HIPERPLASIA PROSTATICA',
       'CARCINOMA INDIFERENCIADO DE CELULAS CLARAS'], dtype=object)

In [68]:
# Se renombran
df['BIOPSIA'] = df['BIOPSIA'].replace('ADENOCARCINOMA GLEASON 6 ',
                                      'ADENOCARCINOMA GLEASON 6')
df['BIOPSIA'] = df['BIOPSIA'].replace('ADENOCARCINOMA GLEASON 7 ',
                                      'ADENOCARCINOMA GLEASON 7')
df['BIOPSIA'] = df['BIOPSIA'].replace('ADENOCARCINOMA GLEASON 8 ',
                                      'ADENOCARCINOMA GLEASON 8')
df['BIOPSIA'] = df['BIOPSIA'].replace('ADENOCARCINOMA GLEASON 9 ',
                                      'ADENOCARCINOMA GLEASON 9')
df['BIOPSIA'] = df['BIOPSIA'].replace('ADENOCARCINOMA GLEASON 10 ',
                                      'ADENOCARCINOMA GLEASON 10')

In [69]:
# Se revisan las categorías de la columna 
df['PATRON DE RESISTENCIA'].unique()

array(['NO', ' AMPI R, CIPRO R, GENTA R, SULFA M R',
       'RESISTENTE A AMPI, CIPRO Y GENTA',
       'RESISTENTE A AMPI, SULFA, CEFADROXILO, CEFUROXIMO, CIPRO Y CEFEPIME, CEFOTAXIMA',
       'MULTI SENSIBLE'], dtype=object)

In [70]:
# Se renombran
df['PATRON DE RESISTENCIA'] = df['PATRON DE RESISTENCIA'].replace(' AMPI R, CIPRO R, GENTA R, SULFA M R', 
                                                                  'R AMPI, R GENTA, R SULFA')
df['PATRON DE RESISTENCIA'] = df['PATRON DE RESISTENCIA'].replace('RESISTENTE A AMPI, CIPRO Y GENTA', 
                                                                  'R AMPI, R CIPRO, R GENTA')
df['PATRON DE RESISTENCIA'] = df['PATRON DE RESISTENCIA'].replace('RESISTENTE A AMPI, SULFA, CEFADROXILO, CEFUROXIMO, CIPRO Y CEFEPIME, CEFOTAXIMA', 
                                                                  'R AMPI, R SULFA, R CEFADROXILO, R CEFUROXIMO, R CIPRO, R CEFEPIME, R CEFOTAXIMA')

In [71]:
# Se revisan nuevamente los tipos de datos y los nulos
utils.verificar_tipo_datos(df)

nombre_campo         tipo_datos  \
0                                                EDAD    [<class 'int'>]   
1                                            DIABETES    [<class 'str'>]   
2                          HOSPITALIZACIÓN ULTIMO MES    [<class 'str'>]   
3                                                 PSA  [<class 'float'>]   
4                                    BIOPSIAS PREVIAS    [<class 'str'>]   
5                                  VOLUMEN PROSTATICO    [<class 'str'>]   
6             ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS    [<class 'str'>]   
7                          NUMERO DE MUESTRAS TOMADAS    [<class 'int'>]   
8                                                 CUP    [<class 'str'>]   
9                   ENF. CRONICA PULMONAR OBSTRUCTIVA    [<class 'str'>]   
10                                            BIOPSIA    [<class 'str'>]   
11  NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA...    [<class 'int'>]   
12                                             FIEBRE    [<class 'str'>]   
13                                                ITU    [<class 'str'>]   
14                                    TIPO DE CULTIVO    [<class 'str'>]   
15                                     AGENTE AISLADO    [<class 'str'>]   
16                              PATRON DE RESISTENCIA    [<class 'str'>]   
17                                    HOSPITALIZACION    [<class 'str'>]   
18                            DIAS HOSPITALIZACION MQ    [<class 'int'>]   
19                           DIAS HOSPITALIZACIÓN UPC    [<class 'int'>]   

    no_nulos_%  nulos_%  nulos  
0        100.0      0.0      0  
1        100.0      0.0      0  
2        100.0      0.0      0  
3        100.0      0.0      0  
4        100.0      0.0      0  
5        100.0      0.0      0  
6        100.0      0.0      0  
7        100.0      0.0      0  
8        100.0      0.0      0  
9        100.0      0.0      0  
10       100.0      0.0      0  
11       100.0      0.0      0  
12       100.0      0.0      0  
13       100.0      0.0      0  
14       100.0      0.0      0  
15       100.0      0.0      0  
16       100.0      0.0      0  
17       100.0      0.0      0  
18       100.0      0.0      0  
19       100.0      0.0      0

Para finalizar, se cambiarán algunos nombres de columnas, dado que algunos son muy largos y se los deja a todos en snake_case.

In [72]:
nombres_nuevos = {
    "EDAD": "edad",
    "DIABETES": "diabetes",
    "HOSPITALIZACIÓN ULTIMO MES": "hospitaliz_ult_mes",
    "PSA": "psa",
    "BIOPSIAS PREVIAS": "biopsias_prev",
    "VOLUMEN PROSTATICO": "vol_prostatico",
    "ANTIBIOTICO UTILIAZADO EN LA PROFILAXIS": "antibiotico_en_profilaxis",
    "NUMERO DE MUESTRAS TOMADAS": "nro_muestras",
    "CUP": "cup",
    "ENF. CRONICA PULMONAR OBSTRUCTIVA": "epoc",
    "BIOPSIA": "biopsia",
    "NUMERO DE DIAS POST BIOPSIA EN QUE SE PRESENTA LA COMPLICACIÓN INFECCIOSA": "nro_dias_con_infecc",
    "FIEBRE": "fiebre",
    "ITU": "itu",
    "TIPO DE CULTIVO": "tipo_cultivo",
    "AGENTE AISLADO": "agente_aislado",
    "PATRON DE RESISTENCIA": "patron_resistencia",
    "HOSPITALIZACION": "hospitalizacion",
    "DIAS HOSPITALIZACION MQ": "dias_hosp_mq",
    "DIAS HOSPITALIZACIÓN UPC": "dias_hosp_upc"
}
df.rename(columns=nombres_nuevos, inplace=True)
df.head(3)

edad diabetes hospitaliz_ult_mes   psa biopsias_prev vol_prostatico  \
0    53       NO                 NO   4.0            NO             SI   
1    56       NO                 NO   7.7            NO             SI   
2    57       NO                 NO  29.0            SI             SI   

        antibiotico_en_profilaxis  nro_muestras cup epoc biopsia  \
0  FLUOROQUINOLONA_AMINOGLICOSIDO            12  NO   NO     NEG   
1  FLUOROQUINOLONA_AMINOGLICOSIDO            12  NO   NO     NEG   
2  FLUOROQUINOLONA_AMINOGLICOSIDO            24  NO   NO     NEG   

   nro_dias_con_infecc fiebre itu tipo_cultivo agente_aislado  \
0                    1     SI  NO           NO             NO   
1                    1     SI  NO           NO             NO   
2                    1     SI  NO  HEMOCULTIVO         E.COLI   

         patron_resistencia hospitalizacion  dias_hosp_mq  dias_hosp_upc  
0                        NO              SI             2              0  
1                        NO              SI             5              0  
2  R AMPI, R GENTA, R SULFA              SI             4              3

# Carga

In [73]:
archivo_limpio = 'BBDD_hospitalizacion_limpio.csv'
df.to_csv(archivo_limpio, index=False, encoding='utf-8')
print(f'Se guardó el archivo {archivo_limpio}')

Se guardó el archivo BBDD_hospitalizacion_limpio.csv
